src/chain/chain_accumulator.rs

```diff
@@ -9,38 +9,54 @@
 use super::{AccumulatingEvent, NetworkEvent, Proof, ProofSet, SectionInfoSigPayload};
 use crate::id::PublicId;
 use log::LogLevel;
-use std::collections::{BTreeMap, BTreeSet, VecDeque};
-use std::mem;
+use std::collections::{BTreeMap, BTreeSet};
+use std::{cmp, mem};
 
 /// The threshold (number of unvoted votes) a node to be considered as unresponsive.
-pub const UNRESPONSIVE_THRESHOLD: usize = 48;
+pub const UNRESPONSIVE_THRESHOLD: usize = 96;
 /// The period (X consensued observations) after which node be considered as unresponsive.
-pub const UNRESPONSIVE_WINDOW: usize = 64;
+pub const UNRESPONSIVE_WINDOW: usize = 128;
+/// The max number of peers that under the responsiveness tracking.
+pub const UNRESPONSIVE_PEERS: usize = 16;
 
-struct VoteStatus {
-    event: AccumulatingEvent,
-    unvoted: BTreeSet<PublicId>,
+#[derive(Default)]
+struct VoteStatuses {
+    unvoted: BTreeMap<PublicId, usize>,
 }
 
-impl VoteStatus {
-    fn new(event: AccumulatingEvent, all_voters: BTreeSet<PublicId>) -> Self {
-        VoteStatus {
-            event: event,
-            unvoted: all_voters,
+impl VoteStatuses {
+    fn add_expectation(
+        &mut self,
+        all_voters: BTreeSet<PublicId>,
+        all_members: &BTreeSet<PublicId>,
+    ) {
+        for id in all_voters {
+            let count = self.unvoted.entry(id).or_insert_with(|| 0);
+            // Avoid punishing for a missed vote long time ago.
+            *count = cmp::min(UNRESPONSIVE_WINDOW, count.saturating_add(1));
+        }
+
+        // Avoid overflowing and pruning too often.
+        if self.unvoted.len() > UNRESPONSIVE_PEERS {
+            self.unvoted = mem::replace(&mut self.unvoted, BTreeMap::new())
+                .into_iter()
+                .filter(|(peer_id, _)| all_members.contains(peer_id))
+                .collect();
         }
     }
 
-    fn add_vote(&mut self, event: &AccumulatingEvent, voter: &PublicId) -> bool {
-        if &self.event == event {
-            let _ = self.unvoted.remove(voter);
-            true
-        } else {
-            false
+    fn add_vote(&mut self, voter: &PublicId) {
+        if let Some(count) = self.unvoted.get_mut(voter) {
```

Comment:
How do you know that vote is for a vote within the window (and not an older one)?

Reply:
I don't know. :(
That's why the `window size` is just approximate.